Change the doc to the new format (#4)
hazelcast-guides/springboot-tomcat-session-replication-on-kubernetes: doc/modules/ROOT/pages/index.adoc

:github-address: https://github.com/hazelcast-guides/springboot-tomcat-session-replication-on-kubernetes

> :templates-url: https://raw.githubusercontent.com/hazelcast-guides/adoc-templates/master

Reviewer: Already have `:templates-url:` below 
```suggestion
```

In `pom.xml` we add the following dependecy for cluster members to find each other in the cluster.
[source,xml]
----
<dependency>
    <groupId>com.hazelcast</groupId>
    <artifactId>hazelcast-kubernetes</artifactId>
    <version>${hazelcast-kubernetes.version}</version>
</dependency>
----

And in the code we change the bean `hazelcastConfig` in `Application.java` to make Hazelcast instances able to discover each other in Kubernetes environment:
[source,java]
----
@Bean
public Config hazelcastConfig() {
    Config config = new Config();
    config.setProperty( "hazelcast.logging.type", "slf4j" );
    config.setInstanceName("hazelcastInstance");
    JoinConfig joinConfig = config.getNetworkConfig().getJoin();
    joinConfig.getMulticastConfig().setEnabled(false);
    joinConfig.getKubernetesConfig().setEnabled(true);
    return config;
}
----

To be able to differentiate which pod responds to our request we add a `podName` variable to the response object. `Application.java` and `CommandController.java` also reads the `podName` as follows:
[source,java]
----
    @Value("#{environment.MY_POD_NAME}")
    private String podName;
----

> We also add the `kubernetes.yaml` file which will create two Micronaut applications and a service to connect them to outer world.

Reviewer: Micronaut?

Author: ```suggestion
We also add the `kubernetes.yaml` file which will create two Spring Boot applications and a service to connect them to outer world.
```

== Create a Docker Image

In order to create the Docker image of the application, we will use the https://github.com/GoogleContainerTools/jib[Jib] tool, and its related Maven plugin. It allows to build containers from Java applications without a Docker file or even changing the pom.xml file. To build the image, run the command below:
[source,bash]
----
$ mvn clean compile com.google.cloud.tools:jib-maven-plugin:1.8.0:dockerBuild
----

> This command will create a docker image named `springboot-tomcat-session-replication-on-kubernetes:0.1.0`. If you want to change the name to push it to some registry, you can run the following instead:

Reviewer: Sorry, I'm a little bit obsessed with this capital issue.
```suggestion
This command will create a Docker image named `springboot-tomcat-session-replication-on-kubernetes:0.1.0`. If you want to change the name to push it to some registry, you can run the following instead:
```

[source,bash]
----
$ mvn clean compile com.google.cloud.tools:jib-maven-plugin:1.8.0:dockerBuild  -Dimage=YOUR-NAME/springboot-tomcat-session-replication-on-kubernetes:0.1.0
----

If you run the above command make sure to change the following part in the `kubernetes.yaml` file.
[source,yaml]
----
containers:
  name: hazelcast-micronaut-container
    image: springboot-tomcat-session-replication-on-kubernetes:0.1.0
    imagePullPolicy: IfNotPresent
    ports:
----

[NOTE]
====
Don't forget to put your docker image in the Kubernetes environment or Docker Hub. 

> If you are using `Minikube` start the Minikube and run the following command before the `mvn clean compile ...` command. 
> [source,bash]
> ----
> $ eval $(minikube docker-env)
> ----
> After this command you will be able to interact with Minikube’s Docker daemon, so the docker image built by `mvn clean compile..` will be accessible by Minikube's Docker daemon. Remember this is only valid for the shell that ran the commmand.
> 
> If you are not using `Minikube` and you ran the  `mvn clean compile ...` command changing the image name with your Docker Hub ID. You can push the Docker image to your registry in Docker Hub with the following commands:
> 
> [source,bash]
> ----
> $ docker login
> $ docker push YOUR-NAME/springboot-tomcat-session-replication-on-kubernetes:0.1.0
> ----
> ====

Author: ```suggestion
If you are using `Minikube` start the Minikube and run the following command before the `mvn clean compile ...` command. 
[source,bash]
----
$ eval $(minikube docker-env)
----
After this command you will be able to interact with Minikube’s Docker daemon, so the docker image built by `mvn clean compile..` will be accessible by Minikube's Docker daemon. Remember this is only valid for the shell that ran the commmand.
====

[NOTE]
====
If you are not using `Minikube` and you ran the  `mvn clean compile ...` command changing the image name with your Docker Hub ID. You can push the Docker image to your registry in Docker Hub with the following commands:

[source,bash]
----
$ docker login
$ docker push YOUR-NAME/springboot-tomcat-session-replication-on-kubernetes:0.1.0
----
====
```

== Configure RBAC

https://github.com/hazelcast/hazelcast-kubernetes[Hazelcast Kubernetes discovery plugin] makes calls to Kubernetes API to provide automatic member discovery. Therefore, it needs to have specific ClusterRole rules granted. You can apply the minimal RBAC configuration (for the `default` service account in the `default` namespace) with the following command.

> By this time make sure you have your kubernetes environment running.

Reviewer: ```suggestion
By this time make sure you have your Kubernetes environment running.
```

[source,bash]
----
$ kubectl apply -f rbac.yaml
----

[NOTE]  
====
- If you use service account other than `default` or namespace other than `default`, you need to modify `rbac.yaml` file.
- If your Kubernetes cluster does not use RBAC, you can skip the "Configure RBAC" step
====

> == Deploy Micronaut Application to Kubernetes

Reviewer: Left over content from Micronaut

Author: ```suggestion
== Deploy The Application to Kubernetes
```

[source,bash]
----
$ kubectl apply -f kubernetes.yaml
----

After this command run the following command to ensure that all pods are running.
[source,bash]
----
$ kubectl get pods
----

== Testing the Application

Launch a curl container inside kubernetes cluster. This will create an interactable shell.

[source,bash]
----

> $ kubectl run curl --rm --image=radial/busyboxplus:curl -i --tty

Reviewer: I wonder why did you change the testing logic and a few others from the original README.md content.

Author: In other guides I used this testing logic, so it is easier to follow each guide if we use the same one. I don't understand what  you mean by "a few others".

Reviewer: I tried this and actually it looks better than the original one - no need for cluster ip at all. What I meant by others were the following curl commands. However, this is OK now.

In `pom.xml` we add the following dependecy for cluster members to find each other in the cluster.

[source,xml]
----
<dependency>
    <groupId>com.hazelcast</groupId>
    <artifactId>hazelcast-kubernetes</artifactId>
    <version>${hazelcast-kubernetes.version}</version>
</dependency>
----

And in the code we change the bean `hazelcastConfig` in `Application.java` to make Hazelcast instances able to discover each other in Kubernetes environment:

[source,java]
----
@Bean
public Config hazelcastConfig() {
    Config config = new Config();
    config.setProperty( "hazelcast.logging.type", "slf4j" );
    config.setInstanceName("hazelcastInstance");
    JoinConfig joinConfig = config.getNetworkConfig().getJoin();
    joinConfig.getMulticastConfig().setEnabled(false);
    joinConfig.getKubernetesConfig().setEnabled(true);
    return config;
}
----

To be able to differentiate which pod responds to our request we add a `podName` variable to the response object. `Application.java` and `CommandController.java` also reads the `podName` as follows:

[source,java]
----
    @Value("#{environment.MY_POD_NAME}")
    private String podName;
----

> We also add the `kubernetes.yaml` file will create two Micronaut applications and a service to connect them to outer world.

Author: ```suggestion
We also add the `kubernetes.yaml` file which will create two Micronaut applications and a service to connect them to outer world.
```

== Create a Docker Image
In order to create the Docker image of the application, we will use the https://github.com/GoogleContainerTools/jib[Jib] tool, and its related Maven plugin. It allows to build containers from Java applications without a Docker file or even changing the pom.xml file. To build the image, run the command below:
[source,bash]
----
$ mvn clean compile com.google.cloud.tools:jib-maven-plugin:1.8.0:dockerBuild
----
This command will create a docker image named `springboot-tomcat-session-replication-on-kubernetes:0.1.0`. If you want an image with your choice of name to push it to some registry, you can run the following instead:
[source,bash]
----
$ mvn clean compile com.google.cloud.tools:jib-maven-plugin:1.8.0:dockerBuild  -Dimage=YOUR-NAME/springboot-tomcat-session-replication-on-kubernetes:0.1.0
----

If you run the above command make sure to change the following part in the `kubernetes.yaml` file.
[source,yaml]
----
containers:
  name: hazelcast-micronaut-container
    image: springboot-tomcat-session-replication-on-kubernetes:0.1.0
    imagePullPolicy: IfNotPresent
    ports:
----

> [NOTE]

Author: Here I don't know how to make it pretty. It would be good if we could split the note to two parts.

Reviewer: Just include another NOTE section. e.g.
```
[NOTE]
====
note1
====

[NOTE]
====
note2
====
```
